File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P538_ConvertBstToGreaterTree.py
# author: tobebetter9527
# since: 2024/3/20 8:30
from typing import Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:
    def convertBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.ans = 0
        self.in_order(root)
        return root

    def in_order(self, root):
        if not root:
            return 0
        self.in_order(root.left)
        root.val = root.val + self.ans
        self.ans = root.val
        self.in_order(root.right)

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P538_ConvertBstToGreaterTree.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P538_ConvertBstToGreaterTree {
    int sum;

    public TreeNode convertBST(TreeNode root) {
        inOrder(root);
        return root;
    }

    private void inOrder(TreeNode root) {
        if (root == null) {
            return;
        }
        inOrder(root.right);
        sum += root.val;
        root.val = sum;
        inOrder(root.left);
    }
}
